src/context.rs

```diff
@@ -268,12 +282,38 @@ impl Context {
         }
     }
 
+    /// Add a record to the reverse lookup cache
+    #[cfg(feature = "dns-relay")]
+    pub fn add_to_reverse_lookup_cache(&self, addr: IpAddr, qname: String) {
+        let mut reverse_lookup_cache = self.reverse_lookup_cache.lock();
+        reverse_lookup_cache.insert(addr, qname);
+    }
+
     /// Check target address ACL (for client)
     pub async fn check_target_bypassed(&self, target: &Address) -> bool {
         match self.config.acl {
             // Proxy everything by default
             None => false,
-            Some(ref a) => a.check_target_bypassed(self, target).await,
+            Some(ref a) => {
+                #[cfg(feature = "dns-relay")]
+                {
+                    match *target {
+                        Address::SocketAddress(ref saddr) => {
+                            // do the reverse lookup in our local cache
+                            let mut reverse_lookup_cache = self.reverse_lookup_cache.lock();
+                            if let Some(qname) = reverse_lookup_cache.get(&saddr.ip()) {
+                                // if a qanme is found
+                                let reverse_addr = Address::DomainNameAddress(qname.to_string(), 0);
+                                if a.check_target_bypassed(self, &reverse_addr).await {
+                                    return true;
+                                }
+                            }
+                        }
+                        _ => (),
+                    }
+                }
+                a.check_target_bypassed(self, target).await
+            }
```

Comment:
`check_target_bypassed` should return `Option<bool>`, and return `None` if nothing matches for hostname.